app/com/arpnetworking/rollups/RollupPartitioningUtils.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Dropbox Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.rollups;
+
+import com.arpnetworking.kairos.client.KairosDbRequestException;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSet;
+
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.Comparator;
+import java.util.Optional;
+
+/**
+ * Utility class to help split {@link RollupDefinition}s into pieces.
+ *
+ * @author Spencer Pearson (spencerpearson@dropbox.com)
+ */
+public final class RollupPartitioningUtils {
+
+    private static final Duration TIMEOUT_HEURISTIC_THRESHOLD = Duration.ofSeconds(30);
+
+    private RollupPartitioningUtils() {}
+
+    /**
+     * Split a {@link RollupDefinition} into a family of cheaper-to-execute {@link RollupDefinition}s which,
+     * run independently, will have an effect equivalent the the input's.
+     *
+     * @param job the {@link RollupDefinition} to split up
+     * @return a set of cheaper jobs which partition the parent's work among them
+     * @throws CannotSplitException if there is no way to split the given job into multiple cheaper ones
+     */
+    public static ImmutableSet<RollupDefinition> splitJob(final RollupDefinition job) throws CannotSplitException {
+        final ImmutableMap<String, String> filterTags = job.getFilterTags();
+        final ImmutableMultimap<String, String> allTags = job.getAllMetricTags();
+        final Optional<String> nextFilterTag = allTags.keySet().stream()
+                .filter(tag -> !filterTags.containsKey(tag)) // already filtered on
+                .filter(tag -> allTags.get(tag).size() > 1) // no point splitting on a single-value tag
+                .max(Comparator.comparing(
```

Comment:
I'm okay with this as a first pass but I think we may want to change to a prioritized list of system tags as the first pass for splitting.  Just something to think about.  The reason I'm unsure about this in the first place is that I know the _os_version tag is likely to surface here for a period of time and it will result in many queries when something like _region might be sufficient to get a meaningful split.

Reply:
Yeah, I share your concerns. A couple alternatives that come to mind:
- "tag whose number of values is closest to 10"
- ["tag whose log-number of values is closest to log-10"](https://www.desmos.com/calculator/vpr78hofys)

...but I thought those were probably too fancy.

My impression is that `_os_version`, like `browser_name` for web-perf metrics, has O(100) values, almost all of which are negligible. So, if we split on `_os_version`, we'll fire off O(100) queries that finish ~immediately (because there's little to no data), plus a handful of expensive queries.

I guess that might get expensive: it looks like querying a big row-key table, which each query needs to do, [takes O(0.1s)](https://grafana.pp.dropbox.com/d/000001001/spencers-playground?tab=queries&panelId=43&edit&fullscreen&orgId=1&from=now-2d&to=now).

So, yeah, I'll change this to "log-number of values closest to log-10."

Reply:
(I'm marking this resolved, but happy to change it if you agree with the "too fancy" criticism and see a better way)